packages/hyperion-autologging/src/ALSurfaceVisibilityPublisher.ts

```diff
@@ -0,0 +1,144 @@
+  ALElementEvent &
+  ALLoggableEvent &
+  Readonly<
+    {
+      surface: string;
+      event: 'component_visibility';
+      intersectionEntry: IntersectionObserverEntry;
+    }
+  >;
+
+export type ALChannelSurfaceVisibilityEvent = Readonly<{
+  al_surface_visibility_event: [ALSurfaceVisibilityEventData],
+}
+>;
+
+export type InitOptions = Types.Options<
+  ALSharedInitOptions<ALChannelSurfaceVisibilityEvent & ALChannelSurfaceMutationEvent>
+>;
+
+
+export function publish(options: InitOptions): void {
+  const { channel } = options;
+
+  // lookup surfaces that are mounted by their root element 
+  const activeSurfaces = new Map<Element | null, ALSurfaceMutationEventData>();
+
+  // We need one observer per threshold
+  const observers = new Map<number, IntersectionObserver>();
+
+  channel.addListener('al_surface_mutation_event', event => {
+    switch (event.event) {
+      case 'mount_component': {
+        if (event.capability?.trackVisibilityThreshold) {
+          const observer = getOrCreateObserver(event.capability.trackVisibilityThreshold);
+          const { element } = event;
+          /**
+           * IntersectionObserver cannot track display:content styles because
+           * these elements don't have their own "box".
+           * So, instead we have to focus on the children of the parent element
+           */
+          if (ALSurfaceUtils.isSurfaceWrapper(element)) {
+            for (let el = element.firstElementChild; el; el = el.nextElementSibling) {
+              observer.observe(el);
+            }
+          } else {
+            observer.observe(element);
+          }
+          activeSurfaces.set(element, event);
+        }
+
+        break;
+      }
+      case 'unmount_component': {
+        if (event.capability?.trackVisibilityThreshold) {
+          const observer = getOrCreateObserver(event.capability.trackVisibilityThreshold);
+          const { element } = event;
+          if (ALSurfaceUtils.isSurfaceWrapper(element)) {
+            for (let el = element.firstElementChild; el; el = el.nextElementSibling) {
+              observer.unobserve(el);
+            }
+          } else {
+            observer.unobserve(element);
+          }
+          activeSurfaces.delete(element);
+        }
+        break;
+      }
+    }
+
+    function getOrCreateObserver(threshold: number): IntersectionObserver {
+      let observer = observers.get(threshold);
+      if (!observer) {
+        observer = new IntersectionObserver(
+          (entries: IntersectionObserverEntry[], _observer: IntersectionObserver) => {
+            /**
+             * Since surface may have many children that we added above, we need to merge
+             * all the entries, however, in most cases we may have only one entry
+             */
+            const visibleSet = new Map<ALSurfaceMutationEventData, IntersectionObserverEntry[]>();
+            for (const entry of entries) {
+              if (entry.isIntersecting) {
```

Comment:
Btw, I'm filtering and opting only for when item becomes visible. There might be value in detecting when it disappear too, in which case, I can remove this condition and pass the value on the event itself. 
What do you think? I'm leaning to do the latter, but the event volume can go up if way too many surfaces turn this feature on.